queue/driver_group_mgo.go

```diff
@@ -0,0 +1,510 @@
+package queue
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// mgoGroupDriver is a type that represents and wraps a queues
+// persistence of jobs *and* locks to a mgoGroupDriver instance.
+type mgoGroupDriver struct {
+	session    *mgo.Session
+	opts       MongoDBOptions
+	name       string
+	group      string
+	instanceID string
+	canceler   context.CancelFunc
+	mu         sync.RWMutex
+	LockManager
+}
+
+// NewMgoGroupDriver creates a driver object given a name, which
+// serves as a prefix for collection names, and a MongoDB connection
+func NewMgoGroupDriver(name string, opts MongoDBOptions, group string) Driver {
+	host, _ := os.Hostname() // nolint
+	return &mgoGroupDriver{
+		name:       name,
+		opts:       opts,
+		group:      group,
+		instanceID: fmt.Sprintf("%s.%s.%s", name, host, uuid.NewV4()),
+	}
+}
+
+// OpenNewMgoGroupDriver constructs and opens a new MongoDB driver instance
+// using the specified session. It is equivalent to calling
+// NewMgoGroup() and calling *MongoDB.Open().
+func OpenNewMgoGroupDriver(ctx context.Context, name string, opts MongoDBOptions, group string, session *mgo.Session) (Driver, error) {
+	d := NewMgoGroupDriver(name, opts, group).(*mgoGroupDriver)
+
+	if err := d.start(ctx, session.Copy()); err != nil {
+		return nil, errors.Wrap(err, "problem starting driver")
+	}
+
+	return d, nil
+}
+
+func (d *mgoGroupDriver) ID() string {
+	d.mu.RLock()
+	defer d.mu.RUnlock()
+
+	return d.instanceID
+}
+
+// Open creates a connection to mgoGroupDriver, and returns an error if
+// there's a problem connecting.
+func (d *mgoGroupDriver) Open(ctx context.Context) error {
+	if d.canceler != nil {
+		return nil
+	}
+
+	session, err := mgo.Dial(d.opts.URI)
+	if err != nil {
+		return errors.Wrapf(err, "problem opening connection to mongodb at '%s", d.opts.URI)
+	}
+
+	return errors.Wrap(d.start(ctx, session), "problem starting driver")
+}
+
+func (d *mgoGroupDriver) start(ctx context.Context, session *mgo.Session) error {
+	d.LockManager = NewLockManager(ctx, d)
+
+	dCtx, cancel := context.WithCancel(ctx)
+	d.canceler = cancel
+
+	session.SetSafe(&mgo.Safe{WMode: "majority"})
+
+	d.mu.Lock()
+	d.session = session
+	d.mu.Unlock()
+
+	startAt := time.Now()
+	go func() {
+		<-dCtx.Done()
+		grip.Info(message.Fields{
```

Comment:
Is this debugging, or does it really matter?

Reply:
debugging